sockeye/data_io.py

```diff
@@ -167,18 +148,16 @@ def get_training_data_iters(source: str, target: str,
     :param max_seq_len_target: Maximum target sequence length.
     :param bucketing: Whether to use bucketing.
     :param bucket_width: Size of buckets.
+    :param sequence_limit: Maximum number of training sequences to read.
     :return: Tuple of (training data iterator, validation data iterator, data config).
     """
     logger.info("Creating train data iterator")
-    train_source_sentences, train_target_sentences = read_parallel_corpus(source,
-                                                                          target,
-                                                                          vocab_source,
-                                                                          vocab_target)
-
-    max_observed_source_len = max((len(s) for s in train_source_sentences if len(s) <= max_seq_len_source), default=0)
-    max_observed_target_len = max((len(t) for t in train_target_sentences if len(t) <= max_seq_len_target), default=0)
+    train_source_sentences = SentenceReader(source, vocab_source, add_bos=False, limit=sequence_limit)
+    train_target_sentences = SentenceReader(target, vocab_target, add_bos=True, limit=sequence_limit)
```

Comment:
Without an explicit call to `read_parallel_corpus`, we could add some comments about when/where we're streaming or storing data.  For instance, we could add to the docstring for `ParallelBucketSentenceIter` that it streams the corpus from disk, stores retained sentences as a NumPy array, etc.  I had to scan through the code to remind myself exactly what the process looks like.

Reply:
I can add a couple of comments, but be aware that this PR does not give us a streaming data iterator. It just streams the int sequences from disk WHILE assigning to buckets. The resulting ndarrays in the buckets are still all in memory.